spec/models/active_shipping_calculator_spec.rb

```diff
@@ -120,39 +103,62 @@
       subject
     end
 
-    it 'should check the cache first before finding rates' do
-      # Since the cache is cleared between the tests, cache.fetch will return a miss,
-      # but by passing a block { Hash.new }, the return value of the block will be
-      # written under the given cache key
-      Rails.cache.fetch(calculator.send(:cache_key, package)) { Hash.new }
-      expect(carrier).not_to receive(:find_rates)
-      subject
+    context "when the cache is warm" do
+      before do
+        # We're stubbing the carrier method because we
+        # need to check that a specific instance of carrier
+        # is receiving or not the function call (otherwise test will
+        # pass but only because carrier we're watching and the carrier
+        # used by the calculator are different)
+        allow(calculator).to receive(:carrier).and_return(carrier)
+      end
+
+      it 'should check the cache first before finding rates' do
+        # Since the cache is cleared between the tests, cache.fetch will return a miss,
+        # but by passing a block { Hash.new }, the return value of the block will be
+        # written under the given cache key so we simulate a warm cache
+        Rails.cache.fetch(calculator.send(:cache_key, package)) { Hash.new }
+        expect(carrier).not_to receive(:find_rates)
+        subject
+      end
     end
 
-    context 'with valid response' do
+    context "when the cache is empty" do
       before do
-        allow(carrier).to receive(:find_rates) { response }
+        # We're stubbing the carrier method because we
+        # need to check that a specific instance of carrier
+        # is receiving or not the function call (otherwise test will
+        # pass but only because carrier we're watching and the carrier
+        # used by the calculator are different)
+        allow(calculator).to receive(:carrier).and_return(carrier)
+        allow(carrier).to receive(:find_rates).and_call_original
```

Comment:
This is needed because since we're stubbing the carrier, the response returned by stubbed find_rates from the instance double is nil so the compute_package call is crashing because it's expecting a response from ActiveShipping